File: game/cave/lost_spring.py
import pygame.mixer

from animation import Animation
from audio_helper import AudioHelper
from cave.cave_resources import CaveResources
from null_state import NullState
from phone_events import PhoneEvents
from state import State
from config import Config


class LostSpring(State):
    def process_events(self, phone_events: PhoneEvents):
        if self.get_frame_index() >= len(CaveResources.hugo_spring):
            return NullState

        if self.one_shot(2.5, "HugoSkydUd"):
            AudioHelper.play(CaveResources.hugo_skyd_ud, self.context.audio_port)

        return None

    def render(self, screen):
        screen.blit(Animation.get_frame(CaveResources.hugo_spring, self.get_frame_index()), (0, 0))

    def on_enter(self) -> None:
        super().on_enter()
        AudioHelper.play(CaveResources.hugo_katapult, self.context.audio_port)

    def on_exit(self) -> None:
        super().on_exit()
        if Config.ARGENTINE_VERSION:
            AudioHelper.fadeout(self.context.cave_bg_music_id, self.context.audio_port, 1)
        else:
            AudioHelper.fadeout(self.context.cave_stemning_id, self.context.audio_port, 1)

File: game-web/src/cave/lostSpring.ts
/** Port of game/cave/lost_spring.py */

import { State, type StateClass } from '../state';
import { Animation } from '../animation';
import { AudioHelper } from '../audioHelper';
import { CaveResources } from './caveResources';
import { NullState } from '../nullState';
import { Config } from '../config';
import type { PhoneEvents } from '../phoneEvents';

export class LostSpring extends State {
  processEvents(_events: PhoneEvents): StateClass | null {
    if (this.getFrameIndex() >= CaveResources.hugoSpring.length) return NullState;

    if (this.oneShot(2.5, 'HugoSkydUd')) AudioHelper.play(CaveResources.hugoSkydUd);

    return null;
  }

  render(ctx: OffscreenCanvasRenderingContext2D): void {
    ctx.drawImage(Animation.getFrame(CaveResources.hugoSpring, this.getFrameIndex()), 0, 0);
  }

  onEnter(): void {
    super.onEnter();
    AudioHelper.play(CaveResources.hugoKatapult);
  }

  onExit(): void {
    super.onExit();
    if (Config.ARGENTINE_VERSION) {
      AudioHelper.fadeout(this.context.caveBgMusicId, 1000);
    } else {
      AudioHelper.fadeout(this.context.caveStemningId, 1000);
    }
  }
}
